### DISCUSSION TITLE
未来会不会回归以按sunshine内应用程序创建不同显示器配置

### DISCUSSION THREAD

**Discussion by cainiao524** [ClassicOldSong/Apollo#48] [Ideas]
现版本链接任意种类的客户端（ipad，win掌机）都会在创建虚拟显示屏是按照设备创建虚拟显示器，而不是像之前按照sunshine内创建的'应用程序‘为名创建虚拟显示屏。
这导致同一个客户机每次串流一直固定一套显示方式，让自带的desktop和virtualscreen选项形同虚设。变成无法从串流客户端主动决定要分屏还是屏蔽主显示器，是否给用支持hdr的使用。
![屏幕截图(12)](https://github.com/user-attachments/assets/9aa694b1-7c8f-4bc1-90df-d2af3010d205)

**Comment by ClassicOldSong** [comment]
启用无头模式才会强制使用虚拟屏。有串流主屏需求的话请关闭无头模式

目前这样设计是有意为之的，虚拟屏的ID跟设备保持一致，目的是让设备可以稳定地当作一块显示器使用，这样不会导致显示器组合过于混乱。可以考虑增加以APP为单位区分显示器ID的模式，但目前APPID本身就不是很稳定，会随着应用改名/改封面图片变化，所以新版解决非Artemis客户端身份验证的问题以后这个行为就废弃了

现阶段依旧可以通过 [MultiMonitorTool](https://www.nirsoft.net/utils/multi_monitor_tool.html) 来实现显示器的开关（24H2可能会有bug，暂时无解）。

**Reply by cainiao524** [reply]
有没有其他方法实现在串流前决定虚拟屏是作为副屏或者独占使用？

**Reply by ClassicOldSong** [reply]
不是很好做，这块逻辑和需求上会存在很多冲突和矛盾，没想清楚就贸然做的话会导致更多问题

如果不通过 MultiMonitorTool 来做的话，目前还有另外两个方法：

1. 设备端多开 Moonlight，每个 Moonlight 的身份是相互独立的 (实际上每次重新配对身份就会刷新)
	- Windows 的话使用便携版的 Moonlight，解压多份即可（不要复制），每份单独配对
	- Android 的话可以用系统自带的应用多开
	- iOS 的话可以同时安装 [官方版](https://apps.apple.com/us/app/moonlight-game-streaming/id1000551566) 和 [砖家版](https://apps.apple.com/us/app/moonlight-zwm/id6590630545)
2. 服务端[多开 Apollo](https://github.com/ClassicOldSong/Apollo/wiki/How-to-start-multiple-instances-of-Apollo) (原理同上)

**Reply by ClassicOldSong** [reply]
HDR 的话目前 Audio/Video tab 下有一个 `遵循客户端 HDR 设置` 的选项，启用后设备配置了HDR的话会自动打开，没有配置的会自动关闭。这个选项不开的话就是保持显示器当前的HDR状态。

**Reply by cainiao524** [reply]
感觉可以留特定appid可选强制决定显示模式切换。毕竟appid名字和背景图一般不会轻易改

**Reply by ClassicOldSong** [reply]
目前不是很想支持，因为显示器组合逻辑这块涉及到的逻辑冲突实在太多了

有需求的话建议还是通过 MultiMonitorTool 配合准备命令来实现，在想要独占的APP上增加准备命令，如：

| 打开时执行命令 | 退出应用时执行命令 |
| ------------ | ------------ |
| MultiMonitorTool.exe /disable \\.\DISPLAY3 | MultiMonitorTool.exe /enable \\.\DISPLAY3 |

虚拟屏会在准备命令执行之前创建好，在退出命令执行完毕以后才断开，所以准备命令可以正常读取显示器组合，也可以正常还原之前的状态。

**Reply by ClassicOldSong** [reply]
想了一下，迟早要把APP的ID给固定下来，所以顺便还是实现了。下一个release就有了

![image](https://github.com/user-attachments/assets/857bb43c-74dc-45c6-aaab-378dbe1e44b6)

**Reply by ClassicOldSong** [reply]
已发布: https://github.com/ClassicOldSong/Apollo/releases/tag/v0.2.3-alpha.1

**Comment by cainiao524** [comment]
感谢大佬

**Comment by cainiao524** [comment]
Windows屎山代码发作。同一个配置应用显示器配置正常，但hdr配置文件每次重连后能用的还会随机变化
ipad和steamdeck oled分别设置hdr配置文件
在ipad设置完hdr配置文件后用重复重连
第一次设置里2个配置文件都能选
第二次只剩ipad的hdr颜色配置文件
第三次同上
4次2个配置文件都出现但默认用deck
5次只有deck配置文件可选
6次同4，
7次默认且只有ipad
8次同4
9次同2

**Comment by cainiao524** [comment]
就算是设备身份作为显示器，也会出现每次重连后可用的hdr配置文件列表出现奇怪的排列组合..

**Reply by ClassicOldSong** [reply]
所以我现在不推荐使用HDR。。。

显示器配置相关的问题在24H2里甚至还变严重了。。。。
